File: ruby/560.subarray-sum-equals-k.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def subarray_sum(nums, k)
  sum_count = Hash.new(0)
  sum_count[0] = 1

  sum = 0
  nums.reduce(0) do |ans, num|
    sum += num
    ans += sum_count[sum - k]
    sum_count[sum] += 1

    ans
  end
end

File: rust/src/560.subarray-sum-equals-k.rs
use std::collections::HashMap;

use crate::Solution;

#[allow(dead_code)]
impl Solution {
    pub fn subarray_sum(nums: Vec<i32>, k: i32) -> i32 {
        let mut sum_count = HashMap::new();
        sum_count.insert(0, 1);

        let (mut sum, mut ans) = (0, 0);

        for num in nums {
            sum += num;

            if let Some(&count) = sum_count.get(&(sum - k)) {
                ans += count;
            }

            *sum_count.entry(sum).or_insert(0) += 1;
        }

        ans
    }
}
